README.md
[![pythonversions](https://img.shields.io/pypi/pyversions/bottlejwt.svg)](https://pypi.python.org/pypi/bottlejwt)
[![Code Climate](https://img.shields.io/codeclimate/github/agalera/bottlejwt.svg)](https://codeclimate.com/github/agalera/bottlejwt)
[![Codecov](https://img.shields.io/codecov/c/github/agalera/bottlejwt.svg)](https://codecov.io/github/agalera/bottlejwt)
[![Travis](https://img.shields.io/travis/agalera/bottlejwt.svg)](https://travis-ci.org/agalera/bottlejwt)

# bottlejwt
JWT plugin for bottle

## installation

Via pip:
```pip install bottlejwt```

Or clone:
```git clone https://github.com/agalera/bottlejwt.git```


## example server:
```python
from bottle import get, install, run
from bottlejwt import JwtPlugin

def validation(auth, auth_value):
    print(auth, auth_value)
    return True

@get("/", auth="any values and types")
def example(auth):  # auth argument is optional!
    return "ok"


install(JwtPlugin(validation, 'secret', algorithm='HS256'))
run(host="0.0.0.0", port="9988")
```

## Test:
```bash
curl http://localhost:9988/?access_token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ
```

## Create Token:
```python
from bottlejwt import JwtPlugin

# is a singleton, you only need to initialize once.
# * If you did install () also work
JwtPlugin(validation, 'secret', algorithm='HS256')

print(JwtPlugin.encode({'name': 'pepito'}))
```


CHANGELOG.md
# CHANGELOG

## 0.0.13 (11-01-2017)
- Fix install package via pip!

## 0.0.12 (06-01-2017)
- Fix tests to python 2.6

## 0.0.11 (05-01-2017)
- Fix compatibility list in pypi
- Add python 3.6 to tests in travis

## 0.0.10 (05-01-2017)
- Add tests
- Travis integration
- Coverage 100%
- Full support python 2.7 and Pypy

## 0.0.9 (05-01-2017)
- Support python 2 (https://github.com/agalera/bottlejwt/pull/1) by stryker250

## 0.0.1 to 0.0.8
- Initial version


test/server.py
from bottle import get, install, app
from bottlejwt import JwtPlugin


def validation(auth, auth_value):
    return auth.get('auth') == auth_value


install(JwtPlugin(validation, 'secret', algorithm='HS256'))


@get("/authenticate", auth=1)
def yes_auth(auth):
    return "ok"


@get("/authenticate_without_argument", auth=1)
def yes_auth_no_argument():
    return "ok"


@get("/without_authenticate")
def no_auth():
    return "ok"


if __name__ == '__main__':
    # run()
    app[0].run(host="0.0.0.0", port="1234")  # pragma: no cover


test/__init__.py
from __future__ import with_statement
from glob import glob
from tools import chdir
import unittest
import sys, os

try:
    import coverage
    coverage.process_startup()
except ImportError:
    pass


if 'fast' in sys.argv:
    sys.stderr.write("Warning: The 'fast' keyword skipps server tests.\n")
    os.environ["TEST_FAST"] = "true"

suite = None
if sys.version_info[:2] in ((2,5), (2,6), (3,1)):
    with chdir(__file__):
        suite = unittest.defaultTestLoader.loadTestsFromNames([n[:-3] for n in glob('test_*.py')])
else:
    suite = unittest.defaultTestLoader.discover(__name__)

def main():
    import bottle
    bottle.debug(True)
    vlevel = 2 if 'verbose' in sys.argv else 0
    result = unittest.TextTestRunner(verbosity=vlevel).run(suite)
    sys.exit((result.errors or result.failures) and 1 or 0)


test/test_bottle_integration.py
from webtest import TestApp
import server
from bottlejwt import JwtPlugin
import unittest


app = TestApp(server.app[0])
pluginjwt = JwtPlugin(False, 'secret', algorithm='HS256')
valid_token = JwtPlugin.encode({'name': 'pepito', 'auth': 1})
invalid_token_1 = JwtPlugin.encode({'name': 'pepito', 'auth': 0})
invalid_token_2 = JwtPlugin.encode({'name': 'pepito'})
invalid_token_3 = "random text"


class TestMethods(unittest.TestCase):
    def test_authenticate_valid_token_url_arg(self):
        r = app.get('/authenticate?access_token=%s' % valid_token)
        assert r.status_code == 200

    def test_authenticate_valid_token_header(self):
        headers1 = {'Authorization': 'bearer %s' % valid_token}
        r1 = app.get('/authenticate', headers=headers1)
        assert r1.status_code == 200

        headers2 = {'Authorization': 'Bearer %s' % valid_token}
        r2 = app.get('/authenticate', headers=headers2)
        assert r2.status_code == 200

    def test_authenticate_valid_token_header_without_type_bearer(self):
        headers = {'Authorization': valid_token}
        r = app.get('/authenticate', headers=headers,
                    expect_errors=True)
        assert r.status_code == 403

        headers = {'Authorization': 'naranjas %s' % valid_token}
        r = app.get('/authenticate', headers=headers,
                    expect_errors=True)
        assert r.status_code == 403

    def test_authenticate_no_token(self):
        r = app.get('/authenticate',
                    expect_errors=True)
        assert r.status_code == 403

    def test_authenticate_invalid_token(self):
        # validation error (auth != 1)
        r1 = app.get('/authenticate?access_token=%s' % invalid_token_1,
                     expect_errors=True)
        assert r1.status_code == 401

        # validation error (no content auth)
        r2 = app.get('/authenticate?access_token=%s' % invalid_token_2,
                     expect_errors=True)
        assert r2.status_code == 401

        # token invalid
        r3 = app.get('/authenticate?access_token=%s' % invalid_token_3,
                     expect_errors=True)
        assert r3.status_code == 403

    def test_authenticate_valid_token_without_argument(self):
        r = app.get('/authenticate_without_argument?access_token=%s' %
                    valid_token)
        assert r.status_code == 200

    def test_without_authenticate(self):
        r = app.get('/without_authenticate')
        assert r.status_code == 200


if __name__ == '__main__':
    unittest.main()


bottle_jwt.py
from bottle import request, HTTPError
import inspect
from functools import wraps
import jwt


class JwtPlugin(object):
    name = 'JwtPlugin'
    api = 2
    keyword = 'auth'

    def __init__(self, validation, key, algorithm="HS512",
                 headers=None, json_decoder=None):
        JwtPlugin.jwt = {
                         'key': key,
                         'algorithm': algorithm,
                         'headers': headers,
                         'json_decoder': json_decoder
        }
        self.validation = validation

    @classmethod
    def encode(self, data):
        return jwt.encode(data, **JwtPlugin.jwt).decode('utf-8')

    def decode(self, data):
        try:
            return jwt.decode(data, **JwtPlugin.jwt)
        except:
            return None

    def get_token(self):
        try:
            token = request.query.get('access_token', None)
            if token:
                return token
            _type, token = request.headers['Authorization'].split(" ")
            if _type != "Bearer":
                return None
            return token
        except:
            return None

    def get_auth(self):
        token = self.get_token()
        if not token:
            raise HTTPError(403, "Forbidden")
        decoded = self.decode(token)
        if decoded is None:
            raise HTTPError(403, "Forbidden, bad token")
        decoded['token'] = token
        return decoded

    def apply(self, callback, route):
        auth_value = route.config.get(self.keyword, None)
        if not auth_value:
            return route.callback

        @wraps(route.callback)
        def wrapper(*args, **kwargs):
            auth = self.get_auth()
            if self.validation(auth, auth_value):
                if self.keyword in inspect.getargspec(route.callback)[0]:
                    kwargs['auth'] = auth
                return route.callback(*args, **kwargs)
            else:
                raise HTTPError(401, "Unauthorized")
        return wrapper



bottlejwt.py
from bottle import request, HTTPError
import inspect
from functools import wraps
import jwt


class JwtPlugin(object):
    name = 'JwtPlugin'
    api = 2
    keyword = 'auth'

    def __init__(self, validation, key, algorithm="HS512",
                 headers=None, json_decoder=None):
        JwtPlugin.jwt = {'key': key,
                         'algorithm': algorithm,
                         'headers': headers,
                         'json_decoder': json_decoder}
        self.validation = validation

    @classmethod
    def encode(self, data):
        kwargs = JwtPlugin.jwt.copy()
        del kwargs['json_decoder']
        r = jwt.encode(data, **kwargs)

        # support python 2.x
        if not isinstance(r, str):
            return r.decode()
        return r

    def decode(self, data):
        try:
            return jwt.decode(data, **JwtPlugin.jwt)
        except:
            return None

    def get_token(self):
        try:
            token = request.query.get('access_token', None)
            if token:
                return token
            _type, token = request.headers['Authorization'].split(" ")
            if _type.lower() != "bearer":
                return None
            return token
        except:
            return None

    def get_auth(self):
        token = self.get_token()
        if not token:
            raise HTTPError(403, "Forbidden")
        decoded = self.decode(token)
        if decoded is None:
            raise HTTPError(403, "Forbidden, bad token")
        decoded['token'] = token
        return decoded

    def apply(self, callback, route):
        auth_value = route.config.get(self.keyword, None)
        if not auth_value:
            return callback

        @wraps(callback)
        def wrapper(*args, **kwargs):
            auth = self.get_auth()
            if self.validation(auth, auth_value):
                sig = inspect.getargspec(callback)
                if self.keyword in sig.args:
                    kwargs[self.keyword] = auth
                return callback(*args, **kwargs)
            else:
                raise HTTPError(401, "Unauthorized")
        return wrapper
